doc/secure_boot/hld_secure_boot.md

```diff
@@ -0,0 +1,383 @@
+![secure_boot_verification_flow](secure_boot_verification_flow.jpg)
+
+##### Run-time verification flow diagram with HW root of trust
+![secure_boot_verification_flow](secure_boot_hw_verification_flow.jpg)
+
+####  1.7.2. <a name='ModuleElementsBreakdown'></a>Module Elements Breakdown
+
+##### Shim Role
+Shim is a simple software package that is designed to work as a first-stage bootloader on UEFI systems. Shim It was developed by a group of Linux developers, so, by using shim we can support secure boot using our custom keys, and not more enforcing to use Microsoft keys.
+
+shim (Debian link)
+
+##### Grub Role
+Grub is a bootloader, that differs from Shim by having the ability to run the kernel and allow us to choose the kernel (i.e: Onie/SONiC) that will be used by using a menu.
+
+##### Vmlinuz Role
+vmlinuz is the name of the Linux kernel executable. vmlinuz is a compressed Linux kernel, and it is capable of loading the operating system into memory so that the computer becomes usable and application programs can be run.
+##### Kernel modules Role
+Kernel modules or KO are pieces of code that can be loaded and unloaded into the kernel upon demand. They extend the functionality of the kernel without the need to reboot the system.
+##### Grub.cfg, initrd
+Initrd - The initial RAM disk (initrd) is an initial root file system that is mounted prior to when the real rootfile system is available. The initrd is bound to the kernel and loaded as part of the kernel boot procedure. The kernel then mounts this initrd as part of the two-stage boot process to load the modules to make the real file systems available and get at the real root file system.
+
+The initrd contains a minimal set of directories and executables to achieve this, such as the insmod tool to install kernel modules into the kernel.
+Regarding SONiC:
+
+In general, the system can boot also without using the initrd (required to build the kernel with some specifics modification), some users comment that in runtime the boot time will be better but others comment the opposite.
+
+SONiC is using the initrd, so we will remain it as in the original repo unsigned. (In order to sign it is required to use a gpg key, but there is not really a use case for that. (Probably future work.)
+
+Grub.cfg - The grub.cfg file is the GRUB configuration file. It is generated by the grub2-mkconfig program using a set of primary configuration files and the grub default file as a source for user configuration specifications. The /boot/grub2/grub.cfg file is first generated during Linux installation and regenerated when a new kernel is installed.
+
+The grub.cfg file contains Bash-like code and a list of installed kernels in an array ordered by sequence of installation. For example, if you have four installed kernels, the most recent kernel will be at index 0, the previous kernel will be at index 1, and the oldest kernel will be at index 3.
```

Comment:
Any suggestions to change the kernel command line options? Some of the options will not set when image generated, but when the image installing in device.

Reply:
I implemented it in the follow flow:
in build time, Im creating a grub.cfg file that it pointing to the final grub.cfg that we be created in the installation phase, by doing this flow, I fix your concerns

PS: It's a common flow in Linux to create this grub.cfg that point to the real grub.cfg

Reply:
@davidpil2002 , the Grub.cfg will be signed, right? How to support the image upgrade by SONiC CLI?
It will add a new menuentry in the grub.cfg.

Reply:
@xumia  we are not signing the grub.cfg, so there is no conflict.
(we are signing the shimx64.efi, grubx64.efi, vmlinuz and Kernel Modules.)